cloud/aws/services/ec2/instances.go

```diff
@@ -39,47 +39,48 @@ func (s *Service) InstanceIfExists(instanceID *string) (*v1alpha1.Instance, erro
 	}
 
 	if len(out.Reservations) > 0 && len(out.Reservations[0].Instances) > 0 {
-		ec2instance := out.Reservations[0].Instances[0]
-		return fromSDKTypeToInstance(ec2instance), nil
+		return fromSDKTypeToInstance(out.Reservations[0].Instances[0]), nil
 	}
 
 	return nil, nil
 }
 
 // CreateInstance runs an ec2 instance.
 func (s *Service) CreateInstance(machine *clusterv1.Machine, config *v1alpha1.AWSMachineProviderConfig, clusterStatus *v1alpha1.AWSClusterProviderStatus) (*v1alpha1.Instance, error) {
-	id := config.AMI.ID
-	if id == nil {
-		id = aws.String(s.defaultAMILookup(clusterStatus.Region))
+
+	input := &v1alpha1.Instance{
+		Type: config.InstanceType,
 	}
 
-	subnets := clusterStatus.Network.Subnets.FilterPublic()
-	if len(subnets) == 0 {
-		return nil, errors.New("need at least one subnet but didn't find any")
+	// Pick image.
+	if config.AMI.ID != nil {
+		input.ImageID = *config.AMI.ID
+	} else {
+		input.ImageID = s.defaultAMILookup(clusterStatus.Region)
 	}
 
-	input := &ec2.RunInstancesInput{
-		ImageId:      id,
-		InstanceType: aws.String(config.InstanceType),
-		MaxCount:     aws.Int64(1),
-		MinCount:     aws.Int64(1),
-		SubnetId:     aws.String(subnets[0].ID),
+	// Pick subnet.
+	if config.Subnet != nil && config.Subnet.ID != nil {
+		input.SubnetID = *config.Subnet.ID
+	} else {
+		sns := clusterStatus.Network.Subnets.FilterPrivate()
+		if len(sns) == 0 {
+			return nil, errors.New("failed to run instance, no subnets available")
+		}
+		input.SubnetID = sns[0].ID
 	}
 
+	// Pick SSH key.
 	if config.KeyName != "" {
 		input.KeyName = aws.String(config.KeyName)
 	}
 
-	reservation, err := s.EC2.RunInstances(input)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to run instances")
-	}
-
-	if len(reservation.Instances) <= 0 {
-		return nil, errors.New("no instance was created after run was called")
+	// Pick instance profile.
```

Comment:
it might be worth noting that there is a default for some fields but not for others